React TabAccordion

arunkumarperumal opened the pull request:
|        Q        |                    A                    |
| --------------- | --------------------------------------- |
| Bug fix?        | no                               |
| New feature?    | no                             |
| New sample?     | yes                             |
| Related issues? | no |

## What's in this Pull Request?
New SPFx web part sample : Tab Accordion Web Part with Property Field Collection Data and tinyMCE for Rich Text Editing
- This Web Part allows users to create content as Tabbed content using Property Field Collection Data and tinyMCE for Rich Text Editing targeted for SharePoint Online. 
- Provides options to view as a Tab or an Accordion. 
- In mobile browser defaults to Accordion view.

hugoabernier commented on samples/react-tabacordion/src/webparts/tabAccordion/TabAccordionWebPart.ts:
```diff
@@ -0,0 +1,205 @@
+      this.s4() + '-' + this.s4() + this.s4() + this.s4();
+  }
+
+  private detectmob(): boolean {
```
Consider using the built-in `isMobileBrowser`  from the `BrowserUtilities` class. 

```typescript
import { BrowserUtilities } from '@microsoft/sp-core-library/lib/BrowserUtilities';

console.log(`isMobileBrowser:`, BrowserUtilities.isMobileBrowser());
```

arunkumarperumal replied:
code is updated

hugoabernier commented on samples/react-tabacordion/src/webparts/tabAccordion/TabAccordionWebPart.ts:
```diff
@@ -0,0 +1,205 @@
+   * @function
+   * Generates a GUID part
+   */
+  private s4(): string {
```
This won't be needed if you use the built in `Guid` functions

arunkumarperumal replied:
Use the built in Guid

hugoabernier commented on samples/react-tabacordion/src/webparts/tabAccordion/components/CAccordion.tsx:
```diff
@@ -0,0 +1,118 @@
+                          init={{                          
+                            content_style: "a {color:rgb(0,120,212) !important}",
+                            plugins: 'link image table lists pageembed',
+                            menubar: 'edit insert format table lists view',  // skip file
```
It would be awesome if you'd use localization instead of static strings. I think that people will most likely want to use your sample as is, and community contributors are most likely to want to add more languages

arunkumarperumal replied:
these are more options or settings for the tiny-mce. I don't think this can be localized.

hugoabernier commented on samples/react-tabacordion/src/webparts/tabAccordion/components/CTab.tsx:
```diff
@@ -0,0 +1,142 @@
+            <div>                 
+              <div className={styles.webpartheader}>
+                <div className={styles.webparttitle}>
+                  <span role="heading" aria-level={2}>{this.props.title}</span>
```
Thank you for thinking about accessibility!

hugoabernier commented on samples/react-tabacordion/src/webparts/tabAccordion/components/utilities/Tab/TabContent.tsx:
```diff
@@ -0,0 +1,50 @@
+        const visibleStyle = this.props.visibleStyle || {};
+        const displayStyle = this.props.isVisible ? visibleStyle : styles.hidden;
+        const disableInlineStyles = this.props.disableInlineStyles;
+        const className = this.props.className || 'tab-content';
```
Avoid static class names. Other web parts on the page could interfere with your web part's styles.

arunkumarperumal replied:
code is updated

hugoabernier commented on samples/react-tabacordion/src/webparts/tabAccordion/components/utilities/Tab/TabContent.tsx:
```diff
@@ -0,0 +1,50 @@
+        const disableInlineStyles = this.props.disableInlineStyles;
+        const className = this.props.className || 'tab-content';
+        const visibleClassName =
+          this.props.visibleClassName || 'tab-content-visible';
```
Again, avoid static class names.

arunkumarperumal replied:
code is updated

hugoabernier commented on samples/react-tabacordion/src/webparts/tabAccordion/components/utilities/Tab/TabLink.tsx:
```diff
@@ -0,0 +1,117 @@
+          ...passedProps
+        } = this.props;
+    
+        const _className = className || 'tab-link';
```
No static class names please

arunkumarperumal replied:
code is updated

hugoabernier commented on samples/react-tabacordion/src/webparts/tabAccordion/components/utilities/Tab/TabLink.tsx:
```diff
@@ -0,0 +1,117 @@
+        } = this.props;
+    
+        const _className = className || 'tab-link';
+        const _activeClassName = activeClassName || 'tab-link-active';
```
No static class names please

arunkumarperumal replied:
code is updated

hugoabernier commented on samples/react-tabacordion/README.md:
```diff
@@ -0,0 +1,117 @@
+
+Solution|Author(s)
+--------|---------
+react-TabAccordion | [Arun Kumar Perumal](https://github.com/arunkumarperumal) - LinkedIn:  <https://www.linkedin.com/in/arunkumarperumal/>
```
No Twitter account? We use your Twitter account to publicize the sample.

arunkumarperumal replied:
I have updated the twitter account details.

hugoabernier reviewed (changes requested):
Please review the recommended changes

hugoabernier reviewed (changes requested):
Please add another image that is not animated. Keep this gif, but add another static preview image.

Also, the folder name is mispelled. it should be `react-accordion` -- but since that already exists, please rename the sample to be easily distinguishable from the other sample. e,g, `react-accordion-with-richtext` or something similar

arunkumarperumal commented on samples/react-tabacordion/src/webparts/tabAccordion/components/utilities/Tab/Tabs.tsx:
```diff
@@ -0,0 +1,170 @@
+import * as React from 'react';
```
Yes,  I have used Fluent UI tabs in my other example of react-faqs. this one was built before UI tabs and wanted to keep the implementation. Should i change to use Fluent UI tabs

arunkumarperumal replied:
I have kept the custom tabs to show the implementation capability.

arunkumarperumal commented on samples/react-tabacordion/src/webparts/tabAccordion/TabAccordionWebPart.ts:
```diff
@@ -0,0 +1,205 @@
+  }
+
+  private detectmob(): boolean {
+    console.log('inside detectmob');
```
removed

arunkumarperumal commented on samples/react-tabacordion/src/webparts/tabAccordion/TabAccordionWebPart.ts:
```diff
@@ -0,0 +1,205 @@
+  private isMobile: boolean;
+
+  // ...
+  protected get propertiesMetadata(): IWebPartPropertiesMetadata {
```
Thanks!

arunkumarperumal commented on samples/react-tabacordion/src/webparts/tabAccordion/components/CAccordion.tsx:
```diff
@@ -0,0 +1,118 @@
+                            },
+                            image_advtab: true,
+                            style_formats: [
+                              {title: 'Headings', items: [
```
updated the headings Heading 1 to h2 and so on

arunkumarperumal commented on samples/react-tabacordion/src/webparts/tabAccordion/components/CAccordion.tsx:
```diff
@@ -0,0 +1,118 @@
+                  <AccordionItemTitle className={styles.accordion__title} id={this.props.guid + '-title-' + index}>
+                  <div className={styles.accordion__arrow} role="presentation" />
+                      <div className={styles["positionRelative"]}>                      
+                         {tab.Title}                          
```
It is actually the Title of the Accordion and hence a custom component. I have used the WebPartTitle separetly

arunkumarperumal commented:
@hugoabernier, 
I have the code with all requested changes. Below are the things that i have updated. 
1. Changed Headings 1 to h2 and accordingly others
2. Use PnPWebPartTitle control. 
3. Used BrowserUtilities for the mobile detection. 
4. Use inbuilt Guid 
5. Removed all static classes. 

I have kept the custom implementation of the tabs, just to show the capability. 

Please let me know if any change required.

hugoabernier commented:
Thank you @arunkumarperumal for the sample, and for the quick turnaround with making the requested changes!!

Would you be interested in demoing this sample on the June 2nd community call? You'd have 10-13 minutes to review the functionality and the key code areas.

arunkumarperumal commented:
Thanks Hugo for accepting the PR. I already have another demo of React FAQs sample on the same day. Can I do the demo of both the webparts in the same call

arunkumarperumal commented:
June 30th demo should be good. Thanks!

hugoabernier commented:
Perfect! I have you confirmed for June 30th!
